homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,160 @@
+"""Denon HEOS Media Player."""
+
+from homeassistant.components.media_player import MediaPlayerDevice
+from homeassistant.components.media_player.const import (
+    DOMAIN, MEDIA_TYPE_MUSIC, SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA, SUPPORT_PREVIOUS_TRACK, SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+from homeassistant.const import STATE_OFF, STATE_PAUSED, STATE_PLAYING
+
+from . import DOMAIN as HEOS_DOMAIN
+
+DEPENDENCIES = ['heos']
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP
+
+PLAY_STATE_TO_STATE = {
+    'play': STATE_PLAYING,
+    'pause': STATE_PAUSED,
+    'stop': STATE_OFF
+}
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    controller = hass.data[HEOS_DOMAIN][DOMAIN]
+    players = controller.get_players()
+    devices = [HeosMediaPlayer(p) for p in players]
+    async_add_devices(devices, True)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, player):
+        """Initialize."""
+        self._player = player
+        self._player.state_change_callback = None
+
+    def _update_state(self):
+        self.async_schedule_update_ha_state()
+
+    async def async_update(self):
+        """Update the player."""
+        self._player.request_update()
+
+    async def async_added_to_hass(self):
```

Comment:
@MartinHjelmare Do we need to also implement `async_will_remove_from_hass` as well?  The component (`heos`) doesn't support unloading in this PR, but the platform (`media_player`) does.

Reply:
No. We're not removing these entities during home assistant lifetime. We also don't seem to need to do anything on remove, like unsubscribing something.

Reply:
> We also don't seem to need to do anything on remove, like unsubscribing something.

We do need to clear the callback so that the component doesn't try to continue updating the state (`self._player.state_change_callback = None`).  This is only a consideration if the component supports unloading in the future.

Reply:
Yes, I understand. But for now we dont remove entities in the platform. And we dont have any special way to remove or disable the state_change_callback. Even if we cant remove an entity, it wont generate any event to update either.

Maybe you have an idea to remove the callback?

Reply:
All that is needed to disconnect the player updates from the home assistant entity is `self._player.state_change_callback = None`.  Based on the code of the player class, it will check for None before trying to invoke the callback.  Given this isn't needed now, I'm going to resolve this conversation.